object_model/artifact.py

```diff
@@ -365,57 +365,179 @@ def pull(self, manifest={}):
+                if literal.__oneByOne__:
+                    old_literal_bindings[literal] = literal.v[0]
+                    literal.v[0] = binding
+                else:
+                    old_literal_bindings[literal] = literal.v
+                    literal.v = binding
+
+
+            # If there are literals that are unbound, then set them to default.
+            if len(bindings) != len(self.__getLiteralsAttached__()):
+                unbound_literals = [literal for literal in self.xp_state.literals if literal not in bindings.keys()]
+                for literal in unbound_literals:
+                    if literal.__oneByOne__:
+                        old_literal_bindings[literal] = literal.v[0]
+                        literal.v[0] = literal.getDefault()
+                    else:
+                        old_literal_bindings[literal] = literal.v
+                        literal.v = literal.getDefault()
+
+            util.activate(self)
 
-                # dst = self.xp_state.versioningDirectory + '/' + "peek0"
-                # copytree(os.getcwd(), dst, True)
+            # Pull appropriate literal and create trial directory
+            self.__pull__()
 
-                # Unpickle and read the file to output.
-                if util.isPickle(dst + "/" + self.loc):
-                    out = func(util.unpickle(dst + '/' + self.loc))
+            # Write the config file
+            config = {}
+            for litName in self.literalNamesAddendum:
+                config[litName] = util.unpickle(self.literalNamesAddendum[litName])
+            with open('.' + experimentName + '.flor', 'w') as fp:
+                json.dump(config, fp)
+
+            dst = tmpexperiment + '/' + "0"
+            copytree(os.getcwd(), dst, True)
+            subtreeMaxed = util.master_pop(self.xp_state.literals)
+
+            # Post processing: restore all the literals back to original state
+            for literal, old_binding in old_literal_bindings.items():
+                if literal.__oneByOne__:
+                    literal.v[0] = old_binding
                 else:
-                    out = []
-                    with open(dst + '/' + self.loc, 'r') as f:
-                        for i in range(head):
-                            out.append(f.readline())
-                    out = func(out)
-
-                os.chdir(original_dir)
-                # [os.remove(i) for i in self.xp_state.ghostFiles] # We want to keep resources intact
-            except Exception as e:
-                print(e)
-                out = e
-        # TODO: add case for peek next.
-
-        # Delete the temporary versioning directory created.
-        try:
-            rmtree(self.xp_state.versioningDirectory)
-        except:
-            pass
-        # Reset versioning directory to original flor.d directory.
-        self.xp_state.versioningDirectory = trueVersioningDir
+                    literal.v = old_binding
+            self.i = 0
```

Comment:
Ok this is a bug. You need to set the `i` of each and every literalAttached to 0. Here, you're setting the `i` of the Artifact, which is wrong.

Reply:
Ah right...fixed!